docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,604 @@
+is stored as a `TemplateVersion` object. This object freezes a copy of the template and also creates
+a list of external references (like git refs, image tags, etc..) that are floating tags by scanning
+all fleet devices, these floating tags are resolved to stable hashes and stored in (or along)
+the `TemplateVersion` object.
+
+When a previously frozen source changes, a new copy of `TemplateVersion` object is created
+with the updated source reference.
+
+If an error happens during the creation of the `TemplateVersion` object, the fleet controller will
+mark a status condition on the fleet object (e.g. missing sources, missing floating tags, etc..),
+and the fleet controller will not proceed with the rollout.
+
+Once a `TemplateVersion` is created the fleet controller starts rolling out that update to devices.
+This update can take a long time (depending on the upgrade strategy). The fleet controller marks
+the latest template version that is being rolled out, as a `fleet-controller/templateVersion`
+annotation in the fleet for observability through the API.
+
+
+When a template is applied to a specific device via the Device API:
+* Floating tags for external references (os image, git refs, etc..) get converted to stable tags (hashes),
+  this gives flightctl the ability to check images being deployed on devices before the device could auto pick-up
+  a floating tag. (i.e. checks verifying the installation of the flightctl agent, and configuration)
+* When later in time a device has labels updated, the device controller will queue the device to be re-evaulated,
+  for example if a change of fleet owner happens because of the label changes.
+* An `annotation` is added to the device in the form of `fleet-controller/templateVersion` to track the latest device
+  template applied to the device.
+
+TemplateVersions are exposed via the API as read-only objects, they can be used to track changes to the fleet
+spec and the external references that were frozen on the devices:
+
+* `/templateversions` - This endpoint enables listing and cleanup of
+                       `TemplateVersion` objects. Allows filtering by labels and owner.
+* `/fleet/{name}/templateversions/{version}` - This endpoint returns a specific `TemplateVersion` object for the fleet.
+
+#### Device management within a fleet
+
+##### Removing a device from a fleet
+
+A device could be stripped from the labels associating it to a specific fleet, in such
+case the `metadata.owner` field will be cleared by the fleet controller, and the device
+spec will need to be manually managed or eventually associated to a different fleet/owner.
+
+##### Changing a device to a different fleet
+
+A device could be moved to a different fleet, in such case the `metadata.owner` field will be updated
+by the fleet controller, and the device will be managed by the new fleet.
+
+This can be performed by modifying device labels to match a new fleet or modifying the fleet's device
+selector.
+
```

Comment:
The fleet controller should apply the latest template version for the new fleet to the device spec.